### PR TITLE
NAB detector that uses timestamps as distal input

### PR DESCRIPTION
**Pull request by mrcslws**
Adds a new "distal_inputs" project.

~~~
Final score for 'distalTimestamps1CellPerColumn' detector on 'reward_low_FP_rate' profile = 57.15
Final score for 'distalTimestamps1CellPerColumn' detector on 'reward_low_FN_rate' profile = 66.10
Final score for 'distalTimestamps1CellPerColumn' detector on 'standard' profile = 62.51
~~~

### CODE REVIEW

**Comment by mrcslws** [comment]
Alright, @BoltzmannBrain, it's ready

File: projects/distal_input/README.md [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# Description
+
+Send timestamps and other contextual inputs into HTM via distal connections.
+
+~~~
```
Comment (BoltzmannBrain) [inline projects/distal_input/README.md]: I would add that this requires NAB, with a link to the repo.

File: projects/distal_input/NAB/config/profiles.json [hunk 2]
```diff
@@ -0,0 +1,26 @@
+{
```
Comment (BoltzmannBrain) [inline projects/distal_input/NAB/config/profiles.json]: Better would be to import this from NAB, and use this directory for anything explicitly different from NAB. Repeated code and documentation can easily lead to inconsistencies and issues.

File: projects/distal_input/NAB/detectors/distal_timestamps_1_cell_per_column.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+from nupic.encoders.date import DateEncoder
+from nupic.encoders.random_distributed_scalar import RandomDistributedScalarEncoder
+
+from nab.detectors.base import AnomalyDetector
+
+from nupic.bindings.algorithms import SpatialPooler
+from nupic.bindings.experimental import ExtendedTemporalMemory
+
+
+class DistalTimestamps1CellPerColumnDetector(AnomalyDetector):
```
Comment (BoltzmannBrain) [inline projects/distal_input/NAB/detectors/distal_timestamps_1_cell_per_column.py]: Add a docstring to give some explanation of what's going on here?

**Comment by BoltzmannBrain** [comment]
@mrcslws is there more you intend on adding to the "distal_inputs" project? I ask because I could also see this setup as nupic.research/projects/NAB, where we can experiment with nupic algorithms and params, and this distal_timestamps_1_cell_per_column would be one such algorithm.

**Comment by mrcslws** [comment]
@BoltzmannBrain Good call. I've now added a NAB project and I'm using that instead.

File: projects/NAB/results/.gitignore [hunk 4]
```diff
@@ -0,0 +1,2 @@
+*
+!final_results.json
```
Comment (BoltzmannBrain) [inline projects/NAB/results/.gitignore]: This didn't seem to work...

File: projects/NAB/README.md [hunk 5]
```diff
@@ -0,0 +1,11 @@
+# Description
+
+Experimental [NAB](https://github.com/numenta/NAB) detectors.
+
+Requires NAB to be installed.
+
+~~~
+python run.py -d null distalTimestamps1CellPerColumn --dataDir ~/nta/NAB/data --windowsFile ~/nta/NAB/labels/combined_windows.json --profilesFile ~/nta/NAB/config/profiles.json --detect --score --normalize
```
Comment (BoltzmannBrain) [inline projects/NAB/README.md]: Does this work? I think the detector names need to be comma-sep: `... -d null,distalTimestamps1CellPerColumn ...`

**Comment by BoltzmannBrain** [comment]
I think @subutai's input would be good on whether he thinks it's appropriate to include a "NAB experiments" directory in nupic.research.

**Comment by subutai** [comment]
> I think @subutai's input would be good on whether he thinks it's appropriate to include a "NAB experiments" directory in nupic.research.

The research repo is mostly freeform, so it is fine to have nab experiments in here. However, I would change the structure to match the current research conventions. All importable code is supposed to be in `htmresearch`, so you could add the detector under `htmresearch/algorithms`, perhaps `htmresearch/algorithms/anomaly_detection/...`. For the project, I would create a more specific name, such as `projects/nab_experiments` rather than `projects/NAB` which could be confused with the NAB repo.

File: projects/NAB/results/.gitignore (hunk 4 above)
Comment (mrcslws) [inline projects/NAB/results/.gitignore]: I think it did? I'm intentionally including `final_results.json` and excluding everything else.

File: projects/NAB/README.md (hunk 5 above)
Comment (mrcslws) [inline projects/NAB/README.md]: Yeah, it works. And it's consistent with the `run.py --help` output `[-d [DETECTORS [DETECTORS ...]]]`

**Comment by mrcslws** [comment]
Sounds good, I changed the project to "nab_experiments", and now the detector is imported as:

~~~python
from htmresearch.algorithms.anomaly_detection.distal_timestamps_1_cell_per_column import (
  DistalTimestamps1CellPerColumnDetector)
~~~

**Comment by BoltzmannBrain** [comment]
👍

**Pull request merged**